### PR TITLE
Simulation

### PR DESCRIPTION
**Pull request by michaelkiper**
## Overview
This PR is to add in the simulation functionality, heavily inspired but [this repo](https://github.jpl.nasa.gov/marchett/EMIT-FC/blob/main/src/simulate.py). In fact, this functionality set should be almost identical, besides from some changes @marchett2 and I had talked about in person.

The main contents of this PR are:
1. src/cover_class/simulation/args.py
      - The arguments for the simulation are now handled in their own file
2. src/cover_class/simulation/simulate.py
      - This file has all the simulation code. It is broken out into a number of smaller functions that should be more understandable. 

## Testing
The testing was conducted with a number of unit tests to observe that the functionality is correct. However, since the data itself is the important characteristic and it's not easily debuggable, I will rely in confirmation by domain experts such as @marchett2 to be able to identify if the simulations are correct.

### CODE REVIEW

**Comment by michaelkiper** [comment]
@pgbrodrick @marchett2 adding you both as reviewers for this PR. Lmk if there is anyone else that should be added since it is more intense.

File: src/cover_class/simulation/simulate.py [hunk 1]
```diff
@@ -1,23 +1,174 @@
...
+        label_result.append(label_subset)
+
+    if spectra_result:
+        spectra = torch.from_numpy(np.concatenate(spectra_result, axis=0)).to(dtype=torch.float32)
+        labels = torch.from_numpy(np.concatenate(label_result, axis=0)).to(dtype=torch.uint8)
+        return FloatTensor(spectra), ByteTensor(labels)
+    return FloatTensor(torch.tensor([], dtype=torch.float32)), ByteTensor(torch.tensor([], dtype=torch.uint8))
+
+
+def _0_init_simulation_state(
```
Comment (pgbrodrick) [inline src/cover_class/simulation/simulate.py:66]: Remove the _0_ prefix - the order is not absolutely static, and this is confusing longer term.  Same is true for the rest of the numbering sequence (I'll refrain from redundant comments).

File: src/cover_class/simulation/simulate.py [hunk 2]
```diff
@@ -1,23 +1,174 @@
...
+
+
+def _3_remove_small_fractions(
+        dirich_fractions: npt.NDArray[np.float32], 
+        mask:             npt.NDArray[np.bool_]
+        
+    ) -> npt.NDArray[np.float32]:
+
+    survivors = dirich_fractions[mask]
+    survivors /= survivors.sum()
```
Comment (pgbrodrick) [inline src/cover_class/simulation/simulate.py:120]: This needs a dimension on it (perhaps it's implied with a broadcasting syntax, but I think it's actually the full sum right now, which is incorrect).  Either way, it's good to make it readable.

File: src/cover_class/simulation/simulate.py [hunk 3]
```diff
@@ -1,23 +1,174 @@
...
+    return dirich_fractions, mask
+
+
+def _3_remove_small_fractions(
+        dirich_fractions: npt.NDArray[np.float32], 
+        mask:             npt.NDArray[np.bool_]
+        
+    ) -> npt.NDArray[np.float32]:
+
+    survivors = dirich_fractions[mask]
```
Comment (pgbrodrick) [inline src/cover_class/simulation/simulate.py:119]: Why separate the two of these?  We're just passing the mask back and forth, forcing both passing in memory and tracking in the main function.  If we're not certain if masking should be applied, just add a flag in the distribution generation...but the min_frac effectively acts as such already.

File: src/cover_class/simulation/simulate.py [hunk 4]
```diff
@@ -1,23 +1,174 @@
...
+    spectra_result: List[npt.NDArray[np.float32]] = []
+    label_result:   List[npt.NDArray[np.uint8]] = []
+    classes, total_n_components, total_n_comp_idxs = _0_init_simulation_state(sim_args)
+
+    alpha = _1_generate_alpha(sim_args.alpha, sim_args.alpha_uniform_low, sim_args.alpha_uniform_high, total_n_components.sum())
+
+    alpha_idx_start = 0
+    for i in range(sim_args.n_iters): # unfortunately, we need to iterate due to how each simulation has different component numbers
+
+        with __ctxblk: # get dirichlet samples
```
Comment (pgbrodrick) [inline src/cover_class/simulation/simulate.py:26]: What's the intended purpose of this down the line?

File: src/cover_class/simulation/simulate.py [hunk 5]
```diff
@@ -1,23 +1,174 @@
...
+
+def run_simulation(sim_args: SimulationArgs, data_args: DataArgs) -> Tuple[FloatTensor, ByteTensor]:
+    spectra_result: List[npt.NDArray[np.float32]] = []
+    label_result:   List[npt.NDArray[np.uint8]] = []
+    classes, total_n_components, total_n_comp_idxs = _0_init_simulation_state(sim_args)
+
+    alpha = _1_generate_alpha(sim_args.alpha, sim_args.alpha_uniform_low, sim_args.alpha_uniform_high, total_n_components.sum())
+
+    alpha_idx_start = 0
+    for i in range(sim_args.n_iters): # unfortunately, we need to iterate due to how each simulation has different component numbers
```
Comment (pgbrodrick) [inline src/cover_class/simulation/simulate.py:24]: Is this true?  Can't you just blank a full matrix, sub-populate the relevant columns per-sample, and then treat as one?  May or may not be more efficient, but if the bulk of time is in functions outside of the dirichlet, it should save.

**Review by pgbrodrick — CHANGES_REQUESTED** [review]
Requested changes in-line.

File: src/cover_class/simulation/simulate.py (hunk 1 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:66]: If they're not static, wouldn't it actually make more sense to have the function name include information about the order of operation? The reason being is that as changes are made to the file, you risk a spaghetti-ification of the file where the control isn't coherently laid out, and this piece of information helps to reduce that risk. 

However, I don't have a strong opinion on this. If you still want that removed, I'll take them out.

File: src/cover_class/simulation/simulate.py (hunk 3 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:119]: > Why separate the two of these?

Logically the generation of the dirichlet distribution and removal of small fractions seem like slightly different operations of different importance given the team meetings. Since the removal of the small fractions is more "debated"/"discussed" topic, I wanted to reflect its seeming importance by creating it's own discrete step in the process. 

> We're just passing the mask back and forth, forcing both passing in memory

I think with how Python works, it is passing through an object reference into the function, not the whole data itself. So you shouldn't be copying allocated arrays with function calls. It should be a very minimal addition to the stack, but I'll do a deep dive into the Pythonic-Numpy memory management and confirm to be sure. 

> If we're not certain if masking should be applied, just add a flag in the distribution generation...but the min_frac effectively acts as such already.

Ya the `min_frac` should be handling that. A user could set it to -1 and have there be no mask. Having another flag would be redundant then.

File: src/cover_class/simulation/simulate.py (hunk 4 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:26]: It just creates a context block that I thought communicates the different operations more clearly than a set of comments would. If you think it's more distracting than helpful, I can remove them.

File: src/cover_class/simulation/simulate.py (hunk 5 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:24]: Possibly? I'll look into how the dirichlet handles "empty" values. I remember I started to look into that but it got tricky. I'll do another pass through and let you know:
A. Is it possible 
B. The time comparison between the 2

File: src/cover_class/simulation/simulate.py (hunk 2 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:120]: The [`dirich_fractions` is a 1D array](https://github.com/emit-sds/cover-class/pull/11/files/fe59f7db1ec309ebc13d125e8a3075571f4db4e6#diff-524cf8604c385d13831dbf399ad3eb2409b21e19ab69fa668885dce896a5bdc1R108) since this just the distribution for a single simulation in the iteration - I'll make it more clearly communicated that you're dividing a 1D array to get it to normalize to 1

There isn't a test for this though in the [test_3_remove_small_fractions](https://github.com/emit-sds/cover-class/pull/11/files/fe59f7db1ec309ebc13d125e8a3075571f4db4e6) unit test - I'll add one.

File: src/cover_class/simulation/simulate.py (hunk 5 above)
Comment (marchett2) [reply src/cover_class/simulation/simulate.py:24]: Is this about getting rid of the loop to generate the mixed spectra? The numpy dirichlet function will give an error if any of the alphas are passed are <= 0: https://numpy.org/doc/2.0/reference/random/generated/numpy.random.dirichlet.html.  The "empty" spots could be set to a very small value, e.g. 1e-10. The number of columns in the matrix will be the total number of components for all the classes, could be large > 200 columns x number of simulated spectra. But worth trying too see if we can generate all sims in one go?

File: src/cover_class/simulation/simulate.py (hunk 2 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:120]: Test was added to make sure that the `dirich_fractions` variable is 1D. And some comments were added to the function for shape clarification. 

Numpy does not natively provide [shape type hints](https://numpy.org/devdocs/reference/typing.html)
There are some alternative options provided in this [stack overflow](https://stackoverflow.com/questions/71109838/numpy-typing-with-specific-shape-and-datatype), however, the only solution that I think would be useful would be to include the other 3rd party [nptyping](https://github.com/ramonhagenaars/nptyping) library. (I don't think using the options from `typing` are good imo)

**Comment by michaelkiper** [comment]
I was thinking through this implementation a little more and came up with some questions/thoughts. @pgbrodrick @marchett2 would both of you be able to provide input on:

For a simulated batch, I think the best implementation for the output labels is to provide a matrix with each row containing the classes used for a particular simulation. The current implementation has a bug int that it [outputs a set of labels of an unaligned shape.](https://github.com/emit-sds/cover-class/pull/11/files#diff-524cf8604c385d13831dbf399ad3eb2409b21e19ab69fa668885dce896a5bdc1R57)

I then have 2 questions:
1. Should the percents used from each class to make the mixture be a provided output as well - specifically referring to the weights from the Dirichlet distribution (I’m not convinced this is the best piece of info to provide to a model though)
2. In providing back a matrix with each row containing the classes used in the mixing, what value should I use as the class label to signify that it’s a ‘null’ entry in the row?
     - I was thinking of using -1
     - A ‘null’ entry is where the Dirichlet fraction for a particular class is lower than the min-frac so that class won’t actually be included, and thus needs to be removed from the output labels

**Comment by marchett2** [comment]
1. Are we using the pipeline to only create the training set for the actual EMIT spectra and the validation set? Or simulated data as test data is going to be included too? If the pipeline is not meant to use any simulated test data, then it's fine if the "true" Dirichlet fraction is not returned. If we still plan to use the pipeline to also create simulated test data, then maybe the option to keep those?  I use the "true" Dirichlet simulated fractions to get performance metrics for the simulated test set. 

2. -1 sounds like a good value for a null class.

**Comment by michaelkiper** [comment]
> 1. Are we using the pipeline to only create the training set for the actual EMIT spectra and the validation set? Or simulated data as test data is going to be included too? If the pipeline is not meant to use any simulated test data, then it's fine if the "true" Dirichlet fraction is not returned. If we still plan to use the pipeline to also create simulated test data, then maybe the option to keep those?  I use the "true" Dirichlet simulated fractions to get performance metrics for the simulated test set.
> 2. -1 sounds like a good value for a null class.

1. I believe the simulation will only create training data in practice. However there's technically nothing stopping it from generating a simulated validation set though. My thoughts were that it may be beneficial to include the fractions for a certain kind of model objective function. However, I think that with enough simulations, a model should learn an identical embedding regardless of the fraction or not. So for now, I'll leave it out.
2. Sounds good, thanks

File: src/cover_class/simulation/simulate.py (hunk 2 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:120]: This should be resolved with the new PyTorch changes.

File: src/cover_class/simulation/simulate.py (hunk 4 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:26]: Removed with new changes

File: src/cover_class/simulation/simulate.py (hunk 5 above)
Comment (michaelkiper) [reply src/cover_class/simulation/simulate.py:24]: Using @marchett2 's idea of setting the alpha to the lowest possible value (in this case the lowest possible positive [IEEE 754 float 32](https://en.wikipedia.org/wiki/IEEE_754) value is 2^-126) worked.

In order to accomplish this, though, the distribution had to be drawn from the PyTorch rng generator, instead of from Numpy's. When I looked into it, there wasn't a way that numpy's Dirichlet distribution allowed to use a matrix of alphas as we intended, but PyTorch did. Thus, it motived the conversion to PyTorch to allow get rid of the iterators and use the matrix-vector ops.

**Comment by michaelkiper** [comment]
@marchett2 @pgbrodrick pinging you both as there's been a major change to this PR branch.

Given the issues with the implementation in Numpy:
1. The Dirichlet distribution did not work with a matrix of alpha values, no matter how I sliced it (still needed iterators)
2. The speed of the numpy operations being fairly slow

I looked at PyTorch as a possible replacement candidate for the Numpy implementation and found it to be a tremendous performance boost:
Simulation size: 1028 | Iterations: 10,000
54.730s for PyTorch versus 465.398s for Numpy (8.5x speed up 🎉)

Additionally, this allows for the DataLoader-Simulator communication to stay in PyTorch without any conversions needing to be made. 

- [x] Verified it works on GPU device
- [x] Speed test comparison - is faster
- [x] Take out iterators - work in matrix/vector operation space

**Comment by michaelkiper** [comment]
Closing this PR due to the "Processing updates" issue where GitHub isn't loading in the commit diff properly. Will open up a new PR as a result

Screenshot 2025-09-09 at 12 54 38 PM
